### DISCUSSION TITLE
eglot discarding useful hover information

### DISCUSSION THREAD

**Discussion by Wilfred** [joaotavora/eglot#1422] [General]
Some LSPs, most notably rust-analyzer, include metadata in the first line of the hover information. As of https://github.com/emacs-mirror/emacs/commit/67c1e6e89cea426e77e7a844376a38a4e7b2ec46, eglot is only showing the first line.

The LSP traffic looks like this:

```
[jsonrpc] e[11:54:47.405] <-- textDocument/hover[4226] {"jsonrpc":"2.0","id":4226,"result":{"contents":{"kind":"markdown","value":"\n```rust\n// size = 16 (0x10), align = 0x8\nlet src: InFileWrapper<HirFileId, &T>\n```"},"range":{"start":{"line":1439,"character":12},"end":{"line":1439,"character":15}}}}
```

but the only text shown in the minibuffer is `// size = 16 (0x10), align = 0x8`, which is the first line of the hover text from LSP. The second line is much more useful, showing the actual type of the local variable I'm hovering on.

Would it be possible to configure this behaviour? I don't see any obvious variables or defcustoms, so I'd need to use advice to replace the entire definition of eglot-hover-eldoc-function.

Alternatively, does it make sense for eglot to truncate this? I know there are already minibuffer settings to control its size, particularly `eldoc-echo-area-use-multiline-p` and `max-mini-window-height`.

**Comment by joaotavora** [comment]
> Alternatively, does it make sense for eglot to truncate this?

I think it does.  The echo area is a very contested outlet and opinions about how it should be used vary wildly.

So I took the conscious decision of making the Eglot Eldoc backend tell the `eldoc-display-in-echo-area-function` that there's a terser version in the `:echo` property. If you type `C-h .` on a piece of documentation you get to the *eldoc* buffer, which is fed via `eldoc-display-in-buffer`, and doesn't have these limitations.

This was a compromise, for sure, taken to appease many users using many servers across many issues over many years :-)

> Would it be possible to configure this behaviour? I don't see any obvious variables or defcustoms, so I'd need to use advice to replace the entire definition of eglot-hover-eldoc-function.

If you need to change this you should rather the functions in `eldoc-display-functions`.  For example you can maybe do something like

```
(defun my/eldoc-I-got-a-big-echo-area-and-dont-mind-it-jumping (docs interactive)
   "Same as `eldoc-display-in-echo-area` ,but ignore the `:echo` cookie".
   (cl-remf docs :echo)
   (eldoc-display-in-echo-area docs interactive))
```

And then replace `eldoc-display-in-echo-area` with `my/eldoc-I-got-a-big-echo-area-and-dont-mind-it-jumping` in the `eldoc-display-functions-defcustom`.

This is independent from `echo-area-use-multiline-p`, and it won't make a difference here.  Even if all backends use the `:echo` cookie, the echo area can still jump (for example if there is more than one backend) regardless of what you put there.  But it will definitely jump _more_ if you ignore the `:echo` cookie.  Do you want that? I suppose you do.

If so, maybe a customization like `(setq echo-area-use-multiline-p 'splurge)` would be better and would also have the practical effect of ignoring the `:echo` cookie.  This would relieve you from doing the slightly larger customization, and you would get _all_ the lines.

If you want this, please propose a patch to bug-gnu-emacs@gnu.org and I think it can be accepted (though I'm not sure if I'm the eldoc maintainer anymore, but it should be fine).  Include a reference to this discussion in the issue.

**Reply by Wilfred** [reply]
Thank you for the helpful and comprehensive response!

How do you feel about allowing customisation of how the eldoc echo cookie is calculated? After playing with some options, I agree a big eldoc value in the minibuffer can be overwhelming, so teaching it to ignore the first line if it starts with `//` would be a huge help.

**Reply by joaotavora** [reply]
Frankly I think you're better off teach rust-analyzer to do that. If it puts that content out there it's for a reason presumably.

But as I said, you can do waht you want in ElDoc already.  Instead of removing the cookie in your special display function, you can mutate it to do that removal.

**Comment by manuel-uberti** [comment]
@Wilfred Could this be of any help/inspiration: https://codeberg.org/mekeor/eglot-signature-eldoc-talkative ?
